pkg/frontend/v2/frontend_test.go

```diff
@@ -332,8 +339,82 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	require.Nil(t, resp)
 
 	ms.checkWithLock(func() {
-		require.Equal(t, 1, len(ms.msgs))
+		require.Len(t, ms.msgs, 1)
+	})
+}
+
+// Test that the front-end shuts down gracefully, i.e. waits for in-flight queries to finish.
+func TestFrontend_GracefulShutdown(t *testing.T) {
+	const (
+		userID = "test"
+		body   = "all good"
+	)
+	ctx := user.InjectOrgID(context.Background(), userID)
+
+	setupDone := make(chan uint64)
+	t.Cleanup(func() {
+		close(setupDone)
 	})
```

Comment:
Why are we closing the channel?

Reply:
I must've misremembered, thinking it's necessary to close channels in order to free them (like files). You're right, apparently it's not necessary, [channels get GC-ed even if not closed](https://stackoverflow.com/a/36613932/265261).